proposals/data-classes.md

```diff
@@ -0,0 +1,241 @@
+
+Support for `readonly` members in object initializers may seem like a small
+feature, but it's important that making a `data` type readonly not come with
+a lot of extra ceremony. The essence of a `data` type is a set of named fields
+and that should stay true regardless of whether or not the fields are 
+`readonly`. It may be tempting for implementation simplicity to try to use
+constructors instead, but this is a design smell that conflates positional
+semantics with `readonly` semantics. Requiring initialization via constructor
+means that field order becomes a public API and requires careful versioning,
+which is not true of mutable `data` types and is a constraint that should
+be irrelevant to `readonly` semantics.
+
+One way to remove dependence on a constructor is simply not make the members
+`readonly` in metadata. The CLR treats `readonly` mostly as guidance -- it
+can easily be overriden using reflection anyway. Most of the safety of
+`readonly` members in C# is not provided by the runtime, but by C# safety
+rules. One way we could enfore compiler rules would be to generate public
+`get`-only properties and make the backing field public and mutable. Object
+initializers would be able to set the properties, but user code wouldn't be
+able to because the backing properties are unspeakable.
+
+One problem with this strategy is `readonly` fields. In that case there is
+no backing field to hide. There are two possible solutions. The first is
+to forbid public `readonly` fields and require properties. The second is
+to make all fields into properties automatically. This is strange because
+we would be generating a property from a field syntax. However, it removes
+what will be a meaningless restriction for the user, only mandated by
+implementation difficulties. Property substition will never be a perfect
+abstraction (reflection will be able to see properties, for example) but the
+solution would probably be able to match user expectations for the vast
+majority of cases. The properties would also be `ref-readonly` returning, so
+even uses of `in` or `ref-readonly` would function as expected.
+
+If data classes contain any `readonly` members that do not have initializers,
+they also do not define a default public constructor like other classes.
+Instead, they define a protected constructor with no arguments, and an
+unspeakable public "initialization" method. This method is called when using
+an object initializer and the compiler verifies that all `readonly` members
+are initialized, or an error is produced.
+
+
+## Extensible data classes (data class subtyping)
+
+Like normal C# classes, data classes are not sealed by default and can be
+inherited from in base classes. If there are any readonly members without
+default initialization in the base class, the subclass is required to
+define a protected constructor. The constructor must assign all readonly
+members of the base class before the constructor ends, or an error is
+produced.
```

Comment:
Do I understand it correctly that because of this, it's not possible to make something like the following code work?

```c#
data class Person
{
    public string Name { get; }
}

data class Student : Person
{
    public int Grade { get; }
}

…

new Student { Name = "Jane Doe", Grade = 1 }
```

Or do these rules only apply to inheritors that are normal classes, not data classes?

Reply:
This should only apply to normal classes. Data classes should inherit the requirements of their base data classes.